Race Condition Between Generated C Language LCM Message Dependencies

# Problem Definition

There are often dependencies between Drake's LCM messages. For example, [lcmt_qp_controller_input](https://github.com/RobotLocomotion/drake/blob/master/drake/lcmtypes/lcmt_qp_controller_input.lcm) depends on [lcmt_zmp_data](https://github.com/RobotLocomotion/drake/blob/master/drake/lcmtypes/lcmt_zmp_data.lcm). If the dependencies are not correctly specified, race conditions can occur where an LCM message is compiled before all of its dependent LCM messages are generated. An example race condition is here: 

https://drake-jenkins.csail.mit.edu/job/linux-gcc-continuous-matlab-open-source/782/console

The relevant part of the log file is copied below:

```
10:34:20 /home/ubuntu/workspace/linux-gcc-continuous-matlab-open-source/build/drake/lcmtypes/drake_lcmt_qp_controller_input.h:18:33: fatal error: drake_lcmt_zmp_data.h: No such file or directory
10:34:20  #include "drake_lcmt_zmp_data.h"
10:34:20                                  ^
10:34:20 compilation terminated.
10:34:21 make[2]: Entering directory `/home/ubuntu/workspace/linux-gcc-continuous-matlab-open-source/build/drake'
10:34:21 [  7%] Generating drake_lcmt_zmp_data.h, drake_lcmt_zmp_data.c, drake/lcmt_zmp_data.hpp, drake/lcmt_zmp_data.py, drake/lcmt_zmp_data.java
```

The above console output snippet indicates that the compiler tried to find `drake_lcmt_zmp_data.h` before it was generated. Looking in [drake-distro/drake/lcmtypes/CMakeLists.txt](https://github.com/RobotLocomotion/drake/blob/master/drake/lcmtypes/CMakeLists.txt), there is apparently code that ensures the C++ code is generated before building compiled:

https://github.com/RobotLocomotion/drake/blob/4937726eca3dd48dea702479f87c54c8031d42d1/drake/lcmtypes/CMakeLists.txt#L49

However, no such logic appears to exist for the generated C code. Maybe similar logic must be provided for the C code?

--- a/drake/lcm/CMakeLists.txt
+++ b/drake/lcm/CMakeLists.txt
@@ -40,10 +40,6 @@ add_library_with_exports(LIB_NAME drakeLcm SOURCE_FILES
     ${lcm_dependent_source_files}
     ${lcm_independent_source_files})
 
-if (lcm_FOUND)
-  add_dependencies(drakeLcm drake_lcmtypes_hpp)
-endif()
-
 target_link_libraries(drakeLcm
     ${lcm_dependent_link_libraries}
     ${lcm_independent_link_libraries})

--- a/drake/lcmtypes/CMakeLists.txt
+++ b/drake/lcmtypes/CMakeLists.txt
@@ -1,15 +1,16 @@
+include(GenerateExportHeader)
 include(${LCM_USE_FILE})
 
-set(python_args PYTHON_SOURCES python_install_sources)
-if(JAVA_FOUND AND TARGET lcm-java)
+if(JAVA_FOUND)
   set(java_args JAVA_SOURCES java_sources)
 endif()
 
 lcm_wrap_types(
-  C_SOURCES sources
+  C_EXPORT drake_lcmtypes
+  C_SOURCES c_sources
   C_HEADERS c_install_headers
   CPP_HEADERS cpp_install_headers
-  ${python_args}
+  PYTHON_SOURCES python_install_sources
   ${java_args}
   lcmt_body_motion_data.lcm
   lcmt_body_wrench_data.lcm
@@ -46,25 +47,28 @@ lcm_wrap_types(
   lcmt_zmp_data.lcm
 )
 
-add_custom_target(drake_lcmtypes_hpp DEPENDS ${cpp_install_headers})
-
-add_library(drake_lcmtypes STATIC ${sources})
-target_link_libraries(drake_lcmtypes lcm)
+lcm_add_library(drake_lcmtypes C ${c_sources} ${c_install_headers})
+generate_export_header(drake_lcmtypes)
 target_include_directories(drake_lcmtypes INTERFACE
-  ${CMAKE_CURRENT_BINARY_DIR})
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>)
 
-add_library(drakeLCMTypes INTERFACE)
-add_dependencies(drakeLCMTypes drake_lcmtypes_hpp)
-target_link_libraries(drakeLCMTypes INTERFACE lcm-coretypes)
+lcm_add_library(drakeLCMTypes CPP ${cpp_install_headers})
 target_include_directories(drakeLCMTypes INTERFACE
-  ${CMAKE_CURRENT_BINARY_DIR})
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}>)
 
-install(FILES ${c_install_headers} DESTINATION include/lcmtypes)
-install(FILES ${cpp_install_headers} DESTINATION include/lcmtypes/drake)
+lcm_install_headers(
+  ${CMAKE_CURRENT_BINARY_DIR}/drake_lcmtypes_export.h
+  ${c_install_headers}
+  ${cpp_install_headers}
+  DESTINATION include/lcmtypes
+)
 
-install(TARGETS drake_lcmtypes
+install(TARGETS drake_lcmtypes drakeLCMTypes
+  EXPORT ${PROJECT_NAME}Targets
+  RUNTIME DESTINATION bin
+  LIBRARY DESTINATION lib${LIB_SUFFIX}
   ARCHIVE DESTINATION lib${LIB_SUFFIX}
-  INCLUDES DESTINATION include
+  INCLUDES DESTINATION include/lcmtypes
 )
 
 lcm_install_python(${python_install_sources})

--- a/drake/systems/lcm/CMakeLists.txt
+++ b/drake/systems/lcm/CMakeLists.txt
@@ -6,7 +6,6 @@ if(lcm_FOUND)
     lcm_translator_dictionary.cc
     lcmt_drake_signal_translator.cc
     serializer.cc)
-  add_dependencies(drakeLCMSystem2 drake_lcmtypes_hpp)
   target_link_libraries(drakeLCMSystem2
     drakeSystemFramework
     drakeLCMTypes

